### ISSUE TITLE
Nonexistent column is accepted as part of constraint definition

### ISSUE THREAD

**Issue by acastiello** [FirebirdSQL/firebird#6936]
I honestly don´t know if this is a bug or if I am missing something from SQL. I created a couple of tables, one for subjects and the other for assignments:

```
 create table subjects (
  subject_id bigint not null constraint subjects_pk primary key,
  subject_name varchar(80) not null,
  subject_start_date date not null
 );

```
` insert into subjects values (1,'subject 1','2021-08-27');`

```
 create table assignments (
  assignment_id bigint not null constraint assignments_pk primary key,
  assignment_mid bigint not null constraint assignment_mk references subjects(subject_id) on update cascade,
  assignment_name varchar(80) not null,
  assignment_date date not null
 );

```
Later, I realized the assignment date should be at least the subject's start date, so I created a constraint.

```
alter table assignments
  add constraint assignment_date_value
    check (assignment_date >= (select assignment_date from subjects where subject_id = assignment_mid)
      );
```

Then I tried:

```
insert into assignments values (1,1,'assignment 1/subject 1','2021-01-01');  -- wrong, should be rejected
insert into assignments values (2,1,'assignment 2/subject 2','2021-08-28');  -- right
```
And both were accepted.

After reviewing the constraint definition (please check it to see if you can spot the bug) I realized that I misspelled and _ASSIGNMENT_DATE_ is not a column of _SUBJECTS_, so for my knowledge, **the whole constraint definition is semantically incorrect and makes no sense but it´s accepted anyway.**

I isolated the _select_ from the _constraint_:

`select assignment_date from subjects where subject_id = 1`

And the error is correctly flagged:

_SQL Error [335544578] [42S22]: Dynamic SQL Error; SQL error code = -206; Column unknown; ASSIGNMENT_DATE; At line 2, column 8 [SQLState:42S22, ISC error code:335544578]_

For curiosity I recreated the table with the constraint embedded on it with identical results

```
create table assignments (
  assignment_id bigint not null constraint assignments_pk primary key,
  assignment_mid bigint not null constraint assignment_mk references subjects(subject_id) on update cascade,
  assignment_name varchar(80) not null,
  assignment_date date not null
    constraint assignment_date_value check (
      assignment_date >= (select ASSIGNMENT_DATE from SUBJECTS where subject_id = assignment_mid)
      )
 );
```

So it was time to fix things:

```
alter table assignments
 drop constraint assignment_date_value;
 
 alter table assignments
  add constraint assignment_date_value
    check (assignment_date >= (select SUBJECT_START_DATE from SUBJECTS where subject_id = assignment_mid)
      );               
```

And everything ran fine:

`insert into assignments values (3,1,'assignment 3/subject 3','2021-01-01');  -- wrong`

_SQL Error [335544558] [23000]: Operation violates CHECK constraint ASSIGNMENT_DATE_VALUE on view or table ASSIGNMENTS; At trigger 'CHECK_155' [SQLState:23000, ISC error code:335544558]_

`insert into assignments values (4,1,'assignment 4/subject 4','2021-08-28');  -- right`

So finally, is this a bug?. Again, not sure if I missed something here. What would FB check with the incorrect definition if it makes any sense?

FB server is V3.0/64 running on W2012 R2

**Comment by hvlad** [comment]
> After reviewing the constraint definition (please check it to see if you can spot the bug) I realized that I misspelled and _ASSIGNMENT_DATE_ is not a column of _SUBJECTS_, so for my knowledge, **the whole constraint definition is semantically incorrect and makes no sense but it´s accepted anyway.**

There is no bug - fields of table where constraint is defined could be used in CHECK CONSTRAINT expressions.
Firebird can't verify semantically correctness of SQL statements, sorry. From the SQL syntax POV there is no error.

Note, while this is possible, but is not recommended to create too complex constrains and\or use another tables there.
More, such constrains will not work reliable in concurrent environment due to transaction isolation.
For complex checks use triggers. 
To make such constraints work you need to use some kind of locking or rely on non-transactional mechanism such as PRIMARY, UNIQUE or FOREIGN KEY constrains.
If you need more info on this - feel free to ask at [fb-support list](https://groups.google.com/g/firebird-support).

**Comment by acastiello** [comment]
Just a last few notes:

> There is no bug - fields of table where constraint is defined could be used in CHECK CONSTRAINT expressions.

Yes, that it is SQL standard and used very much

>Firebird can't verify semantically correctness of SQL statements, sorry.

It does it all the time, doesn´t i? It would be a hell if it didn´t

> From the SQL syntax POV there is no error.

I pointed that. What seems weird is that the check condition when isolated is properly flagged as semantically incorrect, but it is not when embedded in a constraint

Regards

**Comment by mrotteveel** [comment]
@acastiello SQL never checks the semantic correctness (i.e. does the query make sense and produce valid results), only the syntactic correctness (i.e. does it conform to the syntax and are all tables and fields referenced valid). Given `assignment_date` exists in the table you define the check constraint, you can reference it when selecting from a different table. The fact this semantically makes no sense at all doesn't make it syntactically invalid.

The fact the query is invalid in isolation doesn't mean it is invalid in this case, because you're missing the fact it also has access to the context of the row the check constraint applies to. By your reasoning, referencing `assignment_mid` in the query shouldn't be allowed either.

**Comment by acastiello** [comment]
Both you @mrotteveel, and @hvlad are right. I tested in another engine with identical result. The best explanation is the one you point out:

> The fact the query is invalid in isolation doesn't mean it is invalid in this case, because you're missing the fact it also has access to the context of the row the check constraint applies to (imagine as if under the surface there is a join to the current row of `assignments`)

Seems that FB regards it somehow  as (don´like it but...)

```
alter table assignments
  add constraint assignment_date_value
    check (assignment_date >= (select assignment_date from subjects,ASSIGNMENTS where subject_id = assignment_mid)
      );
```

I just want to light he fact that the original constraint should not produce a syntax error but a semantic one. In order to be accepted by a translator, a phrase (or statement) should have lexical, syntactic and semantic correctness, in that order. So these should all be rejected

```
select ç from sometable               -- lexical incorrect. ç does no belong to SQL alphabet
zelect nonExistenField from table -- lexical correct, syntactic incorrect since "zelect" is not a SQL token
select nonExistenField from table -- lexical correct, syntactic correct but semantic incorrect
select assignment_date from subjects where subject_id = 1 -- same as above
alter table assignments
  add constraint assignment_date_value
    check (assignment_date >= (select assignment_date from subjects where subject_id = assignment_mid)
      );          -- My guess it should be as above but as you point, there is some weird join embedded here
```

In any case, as @hvlad shows, sense is up to the programmer:

`select * from subjects where 1=2 -- lexical, syntactic and semantically correct, but makes no sense`

In conclusion, semantic and sense are highly coupled, but are not the same

Thank you all for your time
